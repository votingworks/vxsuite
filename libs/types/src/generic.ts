import check8601 from '@antongolub/iso8601';
import { z } from 'zod';
import { err, ok, Result, wrapException } from './result';

export interface Dictionary<T> {
  [key: string]: Optional<T>;
}
export type Optional<T> = T | undefined;
export interface Provider<T> {
  get(): Promise<T>;
}

/**
 * Parse `value` using `parser`. Note that this takes an object that is already
 * supposed to be of type `T`, not a JSON string. For that, use `safeParseJson`.
 *
 * @returns `Ok` when the parse succeeded, `Err` otherwise.
 */
export function safeParse<T>(
  parser: z.ZodType<T>,
  value: unknown
): Result<T, z.ZodError> {
  const result = parser.safeParse(value);

  if (!result.success) {
    return err(result.error);
  }

  return ok(result.data);
}

/**
 * Parse JSON without throwing an exception if it's malformed. On success the
 * result will be `Ok<unknown>` and you'll need to validate the result yourself.
 * Given malformed JSON, the result will be `Err<SyntaxError>`. Add a parser
 * argument to automatically validate the resulting value after deserializing
 * the JSON.
 */
export function safeParseJson(text: string): Result<unknown, SyntaxError>;
/**
 * Parse JSON and then validate the result with `parser`.
 */
export function safeParseJson<T>(
  text: string,
  parser: z.ZodType<T>
): Result<T, z.ZodError | SyntaxError>;
export function safeParseJson<T>(
  text: string,
  parser?: z.ZodType<T>
): Result<T | unknown, z.ZodError | SyntaxError> {
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch (error) {
    return wrapException(error);
  }

  return parser ? safeParse(parser, parsed) : ok(parsed);
}

/**
 * Parse `value` using `parser` or throw an exception if parsing fails.
 */
export function unsafeParse<T>(parser: z.ZodType<T>, value: unknown): T {
  return safeParse(parser, value).unsafeUnwrap();
}

/**
 * Parser `value` using `parser` or return `undefined` if parsing fails.
 */
export function maybeParse<T>(
  parser: z.ZodType<T>,
  value: unknown
): T | undefined {
  return safeParse(parser, value).ok();
}

declare const Unique: unique symbol;

/**
 * Builds a nominal type based on `Base` with tag `Tag`. This allows creating a
 * type that acts as `Base` but is not assignable from `Base`.
 *
 * @example
 *
 *   type EmailAddress = NewType<string, 'EmailAddress'>;
 *
 *   function parseEmailAddress(emailAddress: string): Optional<EmailAddress> {
 *     if (EMAIL_ADDRESS_PATTERN.test(emailAddress)) {
 *       return emailAddress as EmailAddress;
 *     }
 *     return undefined;
 *   }
 *
 *   declare function sendEmail(to: EmailAddress, subject: string, body: string): Promise<void>;
 *
 *   await sendEmail('admin@example.com', 'Hello', 'World');
 *   //              ^^^^^^^^^^^^^^^^^^^
 *   // Argument of type 'string' is not assignable to parameter of type 'EmailAddress'.
 *
 *   const to = parseEmailAddress('admin@example.com');
 *   assert(to);
 *   await sendEmail(to, 'Hello', 'World');
 */
export type NewType<Base, Tag> = Base & { readonly [Unique]: Tag };

/**
 * Creates a type that is either `T` or a `Promise` wrapping `T`.
 */
export type PromiseOr<T> = T | Promise<T>;

export type Id = string;
export const IdSchema: z.ZodSchema<Id> = z
  .string()
  .nonempty()
  .refine((id) => !id.startsWith('_'), 'IDs may not start with an underscore')
  .refine(
    (id) => /^[-_a-z\d]+$/i.test(id),
    'IDs may only contain letters, numbers, dashes, and underscores'
  );
export const ElectionHash: z.ZodSchema<string> = z
  .string()
  .nonempty()
  .refine(
    (hash) => /^[0-9a-f]*$/i.test(hash),
    'Election hashes must be hex strings containing only 0-9 and a-f'
  );
export const MachineId = z
  .string()
  .nonempty()
  .refine(
    (id) => /^[-A-Z\d]+$/.test(id),
    'Machine IDs may only contain numbers, uppercase letters, and dashes'
  );

export const Iso8601Date = z
  .string()
  .refine(check8601, 'dates must be in ISO8601 format');
export type Iso8601Timestamp = string;
export const Iso8601TimestampSchema = Iso8601Date;
