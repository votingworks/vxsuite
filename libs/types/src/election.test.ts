import * as fc from 'fast-check';
import { sha256 } from 'js-sha256';
import {
  BallotIdSchema,
  getDistrictIdsForPartyId,
  getPartyIdsInBallotStyles,
  safeParse,
  safeParseElection,
  WriteInIdSchema,
} from '.';
import {
  election,
  electionMinimalExhaustive,
  electionWithMsEitherNeither,
  primaryElection,
} from '../test/election';
import {
  CandidateContest,
  CandidateSchema,
  ElectionDefinitionSchema,
  expandEitherNeitherContests,
  getCandidateParties,
  getCandidatePartiesDescription,
  getContestsFromIds,
  getEitherNeitherContests,
  getElectionLocales,
  getPartyAbbreviationByPartyId,
  getPartyFullNameFromBallotStyle,
  getPartyPrimaryAdjectiveFromBallotStyle,
  getPrecinctById,
  getPrecinctIndexById,
  isVotePresent,
  PartyIdSchema,
  PartySchema,
  validateVotes,
  vote,
  withLocale,
  YesNoContest,
} from './election';
import { unsafeParse } from './generic';

test('can build votes from a candidate ID', () => {
  const contests = election.contests.filter((c) => c.id === 'CC');
  const contest = contests[0] as CandidateContest;

  expect(vote(contests, { CC: 'C' })).toEqual({
    CC: [contest.candidates[0]],
  });
});

test('can build votes from an array of candidate IDs', () => {
  const contests = election.contests.filter((c) => c.id === 'CC');
  const contest = contests[0] as CandidateContest;

  expect(
    vote(contests, { [contest.id]: contest.candidates.map((c) => c.id) })
  ).toEqual({
    [contest.id]: contest.candidates,
  });
});

test('can build votes from yesno values', () => {
  expect(vote(election.contests, { YNC: 'yes' })).toEqual({
    YNC: 'yes',
  });
  expect(vote(election.contests, { YNC: 'no' })).toEqual({
    YNC: 'no',
  });
});

test('can build votes from ms-either-neither yesno values', () => {
  expect(
    vote(electionWithMsEitherNeither.contests, {
      MSEN: 'yes',
      MSPO: 'no',
    })
  ).toEqual({
    MSEN: 'yes',
    MSPO: 'no',
  });
});

test('can build votes from a candidate object', () => {
  const contests = election.contests.filter((c) => c.id === 'CC');
  const contest = contests[0] as CandidateContest;
  const candidate = contest.candidates[0];

  expect(vote(contests, { CC: candidate })).toEqual({
    CC: [candidate],
  });
});

test('can get ms-either-neither contests from a list', () => {
  expect(
    getEitherNeitherContests(electionWithMsEitherNeither.contests)
  ).toHaveLength(1);
});

test('can expand ms-either-neither contests into yes no contests', () => {
  const expandedContests = expandEitherNeitherContests(
    electionWithMsEitherNeither.contests
  );
  // There is 1 contest that should have expanded into two.
  expect(expandedContests).toHaveLength(
    1 + electionWithMsEitherNeither.contests.length
  );
  for (let i = 0; i < electionWithMsEitherNeither.contests.length; i += 1) {
    const originalContest = electionWithMsEitherNeither.contests[i];
    if (originalContest.type !== 'ms-either-neither') {
      expect(originalContest).toEqual(expandedContests[i]);
    } else {
      expect(expandedContests[i].type).toBe('yesno');
      expect(expandedContests[i + 1].type).toBe('yesno');
    }
  }
});

test('can expand ms-either-neither contests into yes no contests in a primary', () => {
  const expandedContests = expandEitherNeitherContests(
    electionMinimalExhaustive.contests
  );
  // There is 1 contest that should have expanded into two.
  expect(expandedContests).toHaveLength(
    1 + electionMinimalExhaustive.contests.length
  );
  for (let i = 0; i < electionWithMsEitherNeither.contests.length; i += 1) {
    const originalContest = electionMinimalExhaustive.contests[i];
    if (originalContest.type !== 'ms-either-neither') {
      expect(originalContest).toEqual(expandedContests[i]);
    } else {
      expect(expandedContests[i].type).toBe('yesno');
      expect(expandedContests[i + 1].type).toBe('yesno');
      expect(expandedContests[i].partyId).toEqual(originalContest.partyId);
      expect(expandedContests[i + 1].partyId).toEqual(originalContest.partyId);
    }
  }
});

test('can build votes from a candidates array', () => {
  const contests = election.contests.filter((c) => c.id === 'CC');
  const contest = contests[0] as CandidateContest;
  const { candidates } = contest;

  expect(vote(contests, { CC: candidates })).toEqual({
    CC: candidates,
  });
});

test('vote throws when given a contest id that does not match a contest', () => {
  expect(() => vote([], { nope: 'yes' })).toThrowError('unknown contest nope');
});

test('can get a party primary adjective from ballot style', () => {
  const ballotStyleId = '1D';
  expect(
    getPartyPrimaryAdjectiveFromBallotStyle({
      ballotStyleId,
      election: primaryElection,
    })
  ).toEqual('Democratic');
});

test('can get a party abbreviation by party ID', () => {
  expect(
    getPartyAbbreviationByPartyId({
      partyId: primaryElection.parties[0].id,
      election: { ...primaryElection },
    })
  ).toEqual('D');

  expect(
    getPartyAbbreviationByPartyId({
      partyId: primaryElection.parties[0].id,
      election: { ...primaryElection, parties: [] },
    })
  ).toEqual('');
});

test('can get a party full name from ballot style', () => {
  const ballotStyleId = '1D';
  expect(
    getPartyFullNameFromBallotStyle({
      ballotStyleId,
      election: primaryElection,
    })
  ).toEqual('Democratic Party');
});

test('failing to get a full party name returns an empty string', () => {
  const ballotStyleId = 'DOES_NOT_EXIST';
  expect(
    getPartyFullNameFromBallotStyle({
      ballotStyleId,
      election: primaryElection,
    })
  ).toEqual('');
});

test('special cases party primary adjective transform "Democrat" -> "Democratic"', () => {
  const ballotStyleId = '1D';
  expect(
    getPartyPrimaryAdjectiveFromBallotStyle({
      ballotStyleId,
      election: primaryElection,
    })
  ).toEqual('Democratic');
});

test('defaults to empty string if no party can be found', () => {
  const ballotStyleId = 'bogus';
  expect(
    getPartyPrimaryAdjectiveFromBallotStyle({
      ballotStyleId,
      election: {
        ...election,
        parties: [],
      },
    })
  ).toEqual('');
});

test('getPrecinctById', () => {
  expect(
    getPrecinctById({ election, precinctId: election.precincts[0].id })
  ).toBe(election.precincts[0]);
  expect(getPrecinctById({ election, precinctId: '' })).toBeUndefined();
});

test('getPrecinctIndexById', () => {
  expect(
    getPrecinctIndexById({ election, precinctId: election.precincts[0].id })
  ).toBe(0);
  expect(getPrecinctIndexById({ election, precinctId: '' })).toBe(-1);
});

test('getDistrictIdsForPartyId', () => {
  for (const party of electionMinimalExhaustive.parties) {
    const ballotStylesByParty = electionMinimalExhaustive.ballotStyles.filter(
      ({ partyId }) => party.id === partyId
    );
    for (const districtId of getDistrictIdsForPartyId(
      electionMinimalExhaustive,
      party.id
    )) {
      expect(
        ballotStylesByParty.some(({ districts }) =>
          districts.includes(districtId)
        )
      ).toBe(true);
    }
  }
});

test('getPartyIdsInBallotStyles', () => {
  expect(getPartyIdsInBallotStyles(electionMinimalExhaustive)).toEqual(
    electionMinimalExhaustive.parties.map(({ id }) => id)
  );
});

test('getContestsFromIds', () => {
  expect(getContestsFromIds(electionMinimalExhaustive, [])).toEqual([]);
  expect(
    getContestsFromIds(electionMinimalExhaustive, ['best-animal-mammal'])
  ).toEqual([electionMinimalExhaustive.contests[0]]);
  expect(
    getContestsFromIds(electionMinimalExhaustive, [
      'best-animal-mammal',
      'best-animal-mammal',
    ])
  ).toEqual([electionMinimalExhaustive.contests[0]]);
  expect(() =>
    getContestsFromIds(electionMinimalExhaustive, ['not-a-contest-id'])
  ).toThrowError('Contest not-a-contest-id not found');
});

test('isVotePresent', () => {
  expect(isVotePresent()).toBe(false);
  expect(isVotePresent([])).toBe(false);
  expect(isVotePresent(['yes'])).toBe(true);
  expect(
    isVotePresent([
      election.contests.find(
        (c): c is CandidateContest => c.type === 'candidate'
      )!.candidates[0],
    ])
  ).toBe(true);
});

test('validates votes by checking that contests are present in a given ballot style', () => {
  const ballotStyle = election.ballotStyles[0];

  const yesno = election.contests.find(
    (c): c is YesNoContest => c.type === 'yesno'
  ) as YesNoContest;
  expect(() =>
    validateVotes({
      votes: {
        [yesno.id]: ['yes'],
      },
      ballotStyle,
      election,
    })
  ).not.toThrowError();
  expect(() =>
    validateVotes({ votes: { nope: ['yes'] }, ballotStyle, election })
  ).toThrowError(
    'found a vote with contest id "nope", but no such contest exists in ballot style 1'
  );
});

test('list locales in election definition', () => {
  expect(getElectionLocales(election)).toEqual(['en-US']);
  expect(getElectionLocales(election, 'zh-CN')).toEqual(['zh-CN']);
  expect(getElectionLocales({ ...election, _lang: { 'es-US': {} } })).toEqual([
    'en-US',
    'es-US',
  ]);
});

test('pulls translation keys from the top level object', () => {
  expect(
    withLocale(
      { ...election, _lang: { 'es-US': { title: 'Eleccion General' } } },
      'es-US'
    ).title
  ).toEqual('Eleccion General');
});

test('withLocale ignores undefined keys', () => {
  withLocale(
    {
      ...election,
      ballotStyles: election.ballotStyles.map((bs) => ({
        ...bs,
        partyId: undefined,
        _lang: { 'es-US': {} },
      })),
      _lang: { 'es-US': {} },
    },
    'es-US'
  );
});

test('withLocale ignores missing strings for the locale', () => {
  withLocale(
    {
      ...election,
      ballotStyles: election.ballotStyles.map((bs) => ({
        ...bs,
        partyId: undefined,
        _lang: {},
      })),
      _lang: { 'es-US': {} },
    },
    'es-US'
  );
});

test('pulls translation keys from nested objects', () => {
  expect(
    withLocale(
      {
        ...election,
        parties: [
          unsafeParse(PartySchema, {
            id: 'FED',
            name: 'Federalist',
            abbrev: 'FED',
            fullName: 'Federalist',
            _lang: { 'es-US': { name: 'Federalista' } },
          }),
        ],
        _lang: { 'es-US': {} },
      },
      'es-US'
    ).parties[0].name
  ).toEqual('Federalista');
});

test('treats locale identifier as case-insensitive', () => {
  expect(withLocale(election, 'es-US')).toEqual(withLocale(election, 'eS-Us'));
});

test('passes undefined values through', () => {
  expect(withLocale({ ...election, seal: undefined }, 'es-US')).toHaveProperty(
    'seal',
    undefined
  );
});

test('uses the defaults for anything without a translation', () => {
  expect(withLocale(election, 'en-US').title).toEqual(election.title);
  expect(withLocale(election, 'fr-FR').title).toEqual(election.title);
});

test('trying to vote in the top-level ms-either-neither contest fails', () => {
  expect(() => {
    vote(electionWithMsEitherNeither.contests, {
      '750000015-either-neither': ['yes'],
    });
  }).toThrowError();
});

test('candidate schema', () => {
  // invalid IDs
  safeParse(CandidateSchema, { id: '', name: 'Empty' }).unsafeUnwrapErr();
  safeParse(CandidateSchema, {
    id: '_abc',
    name: 'Starts with underscore',
  }).unsafeUnwrapErr();
  safeParse(CandidateSchema, {
    id: 'write-in',
    name: 'Not a write-in',
    isWriteIn: false,
  }).unsafeUnwrapErr();
  safeParse(CandidateSchema, {
    id: 'some-id',
    name: 'Invalid write-in value',
    isWriteIn: true,
  }).unsafeUnwrapErr();
  // valid IDs
  safeParse(CandidateSchema, {
    id: 'bob-loblaw',
    name: 'Bob Loblaw',
  }).unsafeUnwrap();
  safeParse(CandidateSchema, {
    id: 'write-in',
    name: 'Write-in',
    isWriteIn: true,
  }).unsafeUnwrap();
});

test('write-in ID schema', () => {
  // invalid IDs
  safeParse(WriteInIdSchema, '').unsafeUnwrapErr();
  safeParse(WriteInIdSchema, 'abc').unsafeUnwrapErr();
  safeParse(WriteInIdSchema, '__write-in').unsafeUnwrapErr();
  safeParse(WriteInIdSchema, 'writein').unsafeUnwrapErr();

  // valid IDs
  safeParse(WriteInIdSchema, 'write-in').unsafeUnwrap();
  safeParse(WriteInIdSchema, 'write-in-BOB').unsafeUnwrap();
  safeParse(WriteInIdSchema, 'write-in-1-BOB').unsafeUnwrap();
});

test('ballot ID schema', () => {
  // invalid IDs
  safeParse(BallotIdSchema, '').unsafeUnwrapErr();
  safeParse(BallotIdSchema, '_').unsafeUnwrapErr();

  // valid IDs
  safeParse(BallotIdSchema, 'abc').unsafeUnwrap();
  safeParse(BallotIdSchema, 'abc-123').unsafeUnwrap();
});

test('election schema', () => {
  safeParseElection(electionMinimalExhaustive).unsafeUnwrap();

  fc.assert(
    fc.property(fc.anything(), (value) => {
      safeParseElection(value).unsafeUnwrapErr();
    })
  );

  const candidateContest = election.contests.find(
    (c): c is CandidateContest => c.type === 'candidate'
  )!;
  const yesnoContest = election.contests.find(
    (c): c is YesNoContest => c.type === 'yesno'
  )!;
  const parsedElection = safeParseElection({
    ...election,
    contests: [
      {
        ...candidateContest,
        candidates: [
          ...candidateContest.candidates.map((c) => ({
            ...c,
            partyId: election.parties[0].id,
          })),
          ...candidateContest.candidates.map((c) => ({
            ...c,
            id: `${c.id}-noparty`,
            partyId: undefined,
          })),
        ],
      },
      yesnoContest,
    ],
  }).unsafeUnwrap();

  for (const contest of parsedElection.contests) {
    if (contest.type === 'candidate') {
      for (const candidate of contest.candidates) {
        expect(candidate.partyIds).toEqual(
          candidate.id.endsWith('-noparty')
            ? undefined
            : [election.parties[0].id]
        );
      }
    }
  }
});

test('election scheme results reporting URL', () => {
  expect(() => {
    safeParseElection({
      ...election,
      quickResultsReportingUrl: 'https://results.voting.works/',
    }).unsafeUnwrap();
  }).toThrowError();

  expect(() => {
    safeParseElection({
      ...election,
      quickResultsReportingUrl: 'https://results.voting.works',
    }).unsafeUnwrap();
  }).not.toThrowError();
});

test('loading an election with the old sealURL field', () => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { sealUrl, ...rest } = election;

  expect(
    safeParseElection({
      ...rest,
      // eslint-disable-next-line vx/gts-identifiers
      sealURL: 'https://example.com/seal.png',
    }).unsafeUnwrap().sealUrl
  ).toEqual('https://example.com/seal.png');
});

test('getCandidateParties', () => {
  expect(
    getCandidateParties(election.parties, {
      id: 'bob-loblaw',
      name: 'Bob Loblaw',
    })
  ).toEqual([]);

  expect(
    getCandidateParties(election.parties, {
      id: 'bob-loblaw',
      name: 'Bob Loblaw',
      partyIds: election.parties.map(({ id }) => id),
    })
  ).toEqual(election.parties);

  expect(() =>
    getCandidateParties(election.parties, {
      id: 'bob-loblaw',
      name: 'Bob Loblaw',
      partyIds: [unsafeParse(PartyIdSchema, 'not-a-listed-party')],
    })
  ).toThrowError(/not-a-listed-party/);
});

test('getCandidatePartiesDescription', () => {
  expect(
    getCandidatePartiesDescription(election, {
      id: 'bob-loblaw',
      name: 'Bob Loblaw',
    })
  ).toEqual('');

  expect(
    getCandidatePartiesDescription(election, {
      id: 'bob-loblaw',
      name: 'Bob Loblaw',
      partyIds: election.parties.map(({ id }) => id),
    })
  ).toEqual(election.parties.map(({ name }) => name).join(', '));

  expect(() =>
    getCandidatePartiesDescription(election, {
      id: 'bob-loblaw',
      name: 'Bob Loblaw',
      partyIds: [unsafeParse(PartyIdSchema, 'not-a-listed-party')],
    })
  ).toThrowError(/not-a-listed-party/);
});

test('ElectionDefinitionSchema', () => {
  const electionData = JSON.stringify(election);

  expect(() => {
    unsafeParse(ElectionDefinitionSchema, {
      electionHash: 'abc',
      electionData,
      election: electionMinimalExhaustive,
    });
  }).toThrowError(/hash/);

  expect(
    unsafeParse(ElectionDefinitionSchema, {
      electionHash: sha256(electionData),
      electionData,
      election,
    }).election
  ).toEqual(election);
});
