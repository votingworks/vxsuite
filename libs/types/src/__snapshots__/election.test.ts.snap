// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`safeParseElection shows VXF and CDF parsing errors 1`] = `
[Error: Invalid election definition

VXF error: [
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "undefined",
    "path": [
      "ballotStrings"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "ballotStyles"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "contests"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "undefined",
    "path": [
      "county"
    ],
    "message": "Required"
  },
  {
    "code": "custom",
    "message": "Input not instance of DateWithoutTime",
    "fatal": true,
    "path": [
      "date"
    ]
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "districts"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "id"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "parties"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "precincts"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "state"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "title"
    ],
    "message": "Required"
  }
]

CDF error: [
  {
    "code": "invalid_literal",
    "expected": "BallotDefinition.BallotDefinition",
    "path": [
      "@type"
    ],
    "message": "Invalid literal value, expected \\"BallotDefinition.BallotDefinition\\""
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "BallotFormat"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "Election"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "GeneratedDate"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "GpUnit"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "Issuer"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "IssuerAbbreviation"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "Party"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "SequenceEnd"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "SequenceStart"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "VendorApplicationId"
    ],
    "message": "Required"
  },
  {
    "expected": "'1.0.0'",
    "received": "undefined",
    "code": "invalid_type",
    "path": [
      "Version"
    ],
    "message": "Required"
  }
]]
`;
